### PR TITLE
[MVPN-153] Restart updated client

### PR DESCRIPTION
**Pull request by donce**
- Extend `Monitoring` - `onStatusUp` and `onStatusChangeUp` are different.

### CODE REVIEW

File: src/app/mysterium-client/process-manager.js [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+  }
+
+  async start () {
+    this._startLogging().catch(error => {
+      this._logError(`Starting process logging failed.`, error.message)
+      this._bugReporter.captureErrorException(error)
+    })
+
+    try {
+      await this._startProcess()
```
Comment (shroomist) [inline src/app/mysterium-client/process-manager.js]: side note: this only pings daemon port, doesn't ensure that the process has started fully.

File: src/app/mysterium-client/process-manager.js [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+  }
+
+  _startLocalMonitoring () {
+    this._logInfo(`Starting 'mysterium_client' monitoring`)
+    this._monitoring.start()
+  }
+
+  async _ensureClientVersion () {
+    this._logInfo('Waiting for process to check version')
+    await this._onProcessUp()
```
Comment (shroomist) [inline src/app/mysterium-client/process-manager.js]: side note: it is still possible that `_onProcessUp` fires before we add this listener.
perhaps moving this waiter to `_startProcess()` could fix that

File: src/app/mysterium-client/process-manager.js [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+      this._logInfo('Restarting: killing process')
+      await this._process.kill()
+
+      this._logInfo('Restarting: waiting for process to be down')
+      await this._onProcessDown()
+      this._logInfo('Restarting: process is down, starting it up')
+      await this._process.start()
+      this._logInfo('Restarting: waiting for process to be up')
+      await this._onProcessUp()
+      this._logInfo('Restering: process is up')
```
Comment (shroomist) [inline src/app/mysterium-client/process-manager.js]: ok

File: src/app/mysterium-client/process-manager.js (hunk 2 above)
Comment (shroomist) [reply src/app/mysterium-client/process-manager.js]: or we could wait for `this._monitoring.onStatus((st) => st == true)` here

Comment (donce) [reply src/app/mysterium-client/process-manager.js]: I've added new monitoring methods - now monitoring has `onStatusUp` and `onStatusChangeUp`. This uses `onStatusUp` - it triggers callback on each healtcheck `true`, not only when status changes from `false`.

File: src/app/mysterium-client/process-manager.js (hunk 1 above)
Comment (donce) [reply src/app/mysterium-client/process-manager.js]: Yep, that's why we wait for process up.

File: src/app/mysterium-client/process-manager.js [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+      this._bugReporter.captureErrorException(error)
+    })
+
+    await this._startProcess()
+    this._startLocalMonitoring()
+
+    try {
+      await this._ensureClientVersion()
+    } finally {
+      this._sendRendererHealthcheckUp()
```
Comment (mipo47) [inline src/app/mysterium-client/process-manager.js]: ?? Should we wait until health check answers before continue?

File: src/libraries/mysterium-client/index.js [hunk 5]
```diff
@@ -34,12 +33,20 @@ interface Process {
...
 
+  /**
+   * Stop process gracefully.
+   */
   stop (): Promise<void>,
 
+  /**
+   * Force-kill process.
+   */
+  kill (): Promise<void>,
```
Comment (mipo47) [inline src/libraries/mysterium-client/index.js]: maybe we can combine stop and kill: 
if process doesn't want to stop gracefully then kill it :)

File: src/libraries/util-helpers.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+function parseVersion (version) {
```
Comment (mipo47) [inline src/libraries/util-helpers.js]: move it to VersionCheck

Comment (donce) [reply src/libraries/util-helpers.js]: it can't be Flow.

File: src/libraries/mysterium-client/index.js (hunk 5 above)
Comment (donce) [reply src/libraries/mysterium-client/index.js]: Created a task for that: MVPN-303

**Pull request merged**